### PR TITLE
Decouple the program desc with batch_size in Transformer.

### PR DESCRIPTION
**Pull request by guoshengCS**
Decouple the program desc with batch_size in Transformer. The inference program has been validated to have the same generated sentences for different batch size.
It relies on https://github.com/PaddlePaddle/Paddle/pull/9008 .

### CODE REVIEW

File: fluid/neural_machine_translation/transformer/infer.py [hunk 1]
```diff
@@ -273,6 +301,9 @@ def main():
 
     trg_idx2word = paddle.dataset.wmt16.get_dict(
         "de", dict_size=ModelHyperParams.trg_vocab_size, reverse=True)
+    # Append the <pad> token since the dict provided by dataset.wmt16 does
+    # not include it.
+    trg_idx2word[ModelHyperParams.trg_pad_idx] = "<pad>"
```
Comment (pkuyym) [inline fluid/neural_machine_translation/transformer/infer.py:306]: Please fix this in next PR.

File: fluid/neural_machine_translation/transformer/train.py [hunk 2]
```diff
@@ -138,12 +144,14 @@ def test(exe):
         test_avg_costs = []
         for batch_id, data in enumerate(val_data()):
             if len(data) != TrainTaskConfig.batch_size:
```
Comment (pkuyym) [inline fluid/neural_machine_translation/transformer/train.py]: Please fix this

File: fluid/neural_machine_translation/transformer/infer.py (hunk 1 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/infer.py:306]: Get it.

File: fluid/neural_machine_translation/transformer/train.py (hunk 2 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/train.py]: Done. Refine the validation and use the global statistics.

**Review by pkuyym — APPROVED** [review]
LGTM

**Pull request merged**